pallas-miniprotocols/src/localstate/client.rs

```diff
@@ -0,0 +1,175 @@
+            State::Idle => true,
+            State::Acquired => true,
+            _ => false,
+        }
+    }
+
+    fn assert_agency_is_ours(&self) -> Result<(), Error> {
+        if !self.has_agency() {
+            Err(Error::AgencyIsTheirs)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn assert_agency_is_theirs(&self) -> Result<(), Error> {
+        if self.has_agency() {
+            Err(Error::AgencyIsOurs)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn assert_outbound_state(&self, msg: &Message<Q>) -> Result<(), Error> {
+        match (&self.0, msg) {
+            (State::Idle, Message::Acquire(_)) => Ok(()),
+            (State::Idle, Message::Done) => Ok(()),
+            (State::Acquired, Message::Query(_)) => Ok(()),
+            (State::Acquired, Message::Release) => Ok(()),
+            _ => Err(Error::InvalidOutbound),
+        }
+    }
+
+    fn assert_inbound_state(&self, msg: &Message<Q>) -> Result<(), Error> {
+        match (&self.0, msg) {
+            (State::Acquiring, Message::Acquired) => Ok(()),
+            (State::Acquiring, Message::Failure(_)) => Ok(()),
+            (State::Querying, Message::Result(_)) => Ok(()),
+            _ => Err(Error::InvalidInbound),
+        }
+    }
+
+    pub fn send_message(&mut self, msg: &Message<Q>) -> Result<(), Error> {
+        self.assert_agency_is_ours()?;
+        self.assert_outbound_state(msg)?;
+        self.1.send_msg_chunks(msg).map_err(Error::ChannelError)?;
+
+        Ok(())
+    }
+
+    pub fn recv_message(&mut self) -> Result<Message<Q>, Error> {
+        self.assert_agency_is_theirs()?;
+        let msg = self.1.recv_full_msg().map_err(Error::ChannelError)?;
+        self.assert_inbound_state(&msg)?;
+
+        Ok(msg)
+    }
+
+    pub fn send_acquire(&mut self, point: Option<Point>) -> Result<(), Error> {
+        let msg = Message::<Q>::Acquire(point);
+        self.send_message(&msg)?;
+        self.0 = State::Acquiring;
+
+        Ok(())
+    }
+
+    pub fn recv_acquiring(&mut self) -> Result<(), Error> {
+        match self.recv_message()? {
+            Message::Acquired => {
+                self.0 = State::Acquired;
+                Ok(())
+            }
+            Message::Failure(x) => {
+                self.0 = State::Idle;
+                Err(Error::from(x))
+            }
+            _ => Err(Error::InvalidInbound),
+        }
+    }
+
+    pub fn acquire(&mut self, point: Option<Point>) -> Result<(), Error> {
+        self.send_acquire(point)?;
+        self.recv_acquiring()
+    }
+
+    pub fn send_query(&mut self, request: Q::Request) -> Result<(), Error> {
+        let msg = Message::<Q>::Query(request);
+        self.send_message(&msg)?;
+        self.0 = State::Querying;
+
+        Ok(())
+    }
+
+    pub fn recv_querying(&mut self) -> Result<Q::Response, Error> {
+        match self.recv_message()? {
+            Message::Result(x) => {
+                self.0 = State::Idle;
```

Comment:
Think this should be `State::Acquired` unless this was intentional

Reply:
my bad